### ISSUE TITLE
Data.Sequence: traverseWithIndex much slower than sequenceA . mapWithIndex

### ISSUE THREAD

**Issue by arybczak** [haskell/containers#603]
Consider the following program:

```haskell
module Main where

import System.Environment
import qualified Data.Sequence as S

main :: IO ()
main = do
  [sf, sx] <- getArgs
  let x = read sx :: Integer
      f = case sf of
        "map"      -> \g -> sequenceA . S.mapWithIndex g
        "traverse" -> \g -> S.traverseWithIndex g
  putStrLn . show . sum =<< f
    (\i a -> pure $ fromIntegral i + 2*a)
    (S.fromList [1..x])
```

When compiled with `-O`:

```bash
unknown@electronics haskell $ time ./seqtest traverse 1000000
1500000500000

real    0m0,815s
user    0m0,742s
sys     0m0,070s
unknown@electronics haskell $ time ./seqtest map 1000000
1500000500000

real    0m0,398s
user    0m0,359s
sys     0m0,039s
```

This is with GHC 8.6.3 and containers 0.6.0.1.

**Comment by treeowl** [comment]
That's a bit surprising. `traverse` and `traverseWithIndex` are very similar. I wonder if the specialization works out differently. `traverseWithIndex` has some associated rules that are mysteriously missing from `traverse`. Maybe the `NOINLINE [1]` pragmas need to be `INLINABLE [1]` instead? That seems pretty plausible.

**Comment by treeowl** [comment]
While we're at it: there's some really ugly hand-specialization to get around the GHC behavior of only specializing when all instances are known. But I seem vaguely to recall that that restriction may have been removed in recent versions. If so, it would be nice to clean up the code.

**Comment by treeowl** [comment]
@oisdk, do you think you could have a look at this?

**Comment by oisdk** [comment]
I'll try to get to it over the next few days!

**Comment by treeowl** [comment]
Thanks!

**Comment by treeowl** [comment]
Fixed in #623.
